### PR TITLE
soc: stm32: Add STM32L496XE support

### PR DESCRIPTION
**Pull request by gon1332**
Added support for the [STM32L496XE series](https://www.st.com/en/microcontrollers-microprocessors/stm32l496re.html) with 512KB internal flash and 256KB SRAM of contiguous addresses.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

### CODE REVIEW

**Comment by gon1332** [comment]
@nikooiko @galazari

**Comment by psidhu** [comment]
> 512MB

That's a lot of space :smile:

**Comment by gon1332** [comment]
@psidhu Oops. Thanks. That's a hell of an upgrade. I've changed the PR description. I'll change the commit message as soon as possible.

**Comment by gon1332** [comment]
@psidhu @erwango I fixed the commit message. Could you check?

**Comment by psidhu** [comment]
@gon1332 Commit message LGTM.

**Review by erwango — DISMISSED** [review]

File: dts/arm/st/l4/stm32l496Xe.dtsi [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * Copyright (c) 2019 Centaur Analytics, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <mem.h>
+#include <st/l4/stm32l496.dtsi>
+
+/ {
+	sram0: memory@20000000 {
+		reg = <0x20000000 DT_SIZE_K(320)>;
```
Comment (erwango) [inline dts/arm/st/l4/stm32l496Xe.dtsi]: SRAM size is actually 256KB.
We need to take into account that 64Kb RAM are located into SRAM2 at a non contiguous address.

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/l4/stm32l496Xe.dtsi (hunk 1 above)
Comment (gon1332) [reply dts/arm/st/l4/stm32l496Xe.dtsi]: @erwango Done.

**Review by erwango — APPROVED** [review]

**Comment by gon1332** [comment]
@erwango I'd like to know if this PR is valid for Zephyr 1.14. Should I propose it now since it's not yet merged or on a new PR?

**Comment by erwango** [comment]
> @erwango I'd like to know if this PR is valid for Zephyr 1.14. Should I propose it now since it's not yet merged or on a new PR?

@gon1332, I don't think it is. Aim is to fix bugs, not to add new stuff (new SoC in this case), but we can ask @nashif.

**Comment by ioannisg** [comment]
Is there going to be a board for this SOC in the tree, @gon1332 ?

**Comment by gon1332** [comment]
@ioannisg AFAIC, there is no commercial board available. We only have our custom boards based on this SOC. So the answer is no, until there is something commercial.

**Comment by erwango** [comment]
@gon1332 , due to use of `CONFIG_SOC_STM32L496XG` in https://github.com/zephyrproject-rtos/zephyr/pull/16813, I think you'll need to move `SOC_STM32L496XG` to `SOC_STM32L496XX`
so it works for both variants.
Change is safe, more details in https://github.com/zephyrproject-rtos/zephyr/issues/16733.

**Comment by gon1332** [comment]
@erwango Done

**Review by nashif — APPROVED** [review]

**Pull request merged**